zenodo/modules/stats/exporters.py

```diff
@@ -0,0 +1,135 @@
+from invenio_search import current_search_client
+
+from zenodo.modules.stats.utils import chunkify, fetch_record
+
+
+class PiwikExporter:
+    """Events exporter."""
+
+    def run(self, start_date=None, end_date=None, update_bookmark=True):
+        """Run export job."""
+
+        start_date = dateutil_parse(start_date) if start_date else None
+        end_date = dateutil_parse(end_date) if end_date else None
+
+        if start_date is None:
+            bookmark = current_cache.get('piwik_export:bookmark')
+            start_date = dateutil_parse(bookmark) if bookmark else None
+
+        time_range = {}
+        if start_date is not None:
+            time_range['gte'] = start_date.replace(microsecond=0).isoformat()
+        if end_date is not None:
+            time_range['lte'] = end_date.replace(microsecond=0).isoformat()
+
+        events = Search(
+            using=current_search_client,
+            index='events-stats-*'
+        ).filter(
+            'range', timestamp=time_range
+        ).sort(
+            {'timestamp': {'order': 'asc'}}
+        ).params(preserve_order=True).scan()
+
+        url = current_app.config['ZENODO_STATS_PIWIK_EXPORTER'].get('url', None)
+        token_auth = current_app.config['ZENODO_STATS_PIWIK_EXPORTER'] \
+            .get('token_auth', None)
+
+        chunk_size = 85
+        for event_chunk in chunkify(events, chunk_size):
+            query_strings = []
+            for event in event_chunk:
+                query_string = self._build_query_string(event)
+                query_strings.append(query_string)
+
+            payload = {
+                'requests': query_strings,
+                'token_auth': token_auth
+            }
+
+            res = requests.post(url, json=payload)
+
+            # Failure: not 200 or not "success"
+            content = json.loads(res.content)
+            if res.status_code == 200 and content.get('status') == 'success':
+                if content.get('invalid') != 0:
+                    msg = 'Invalid events in Piwik export.'
+                    extra = {
+                        'begin_event': event_chunk[0].timestamp,
+                        'end_event': event_chunk[len(event_chunk)-1].timestamp,
+                        'invalid_events': content.get('invalid')
+                    }
+                    current_app.logger.warning(msg, extra=extra)
+                elif update_bookmark is True:
+                    current_cache.set('piwik_export:bookmark',
+                                      event_chunk[len(event_chunk)-1].timestamp,
+                                      timeout=-1)
+            else:
+                msg = 'Invalid events in Piwik export.'
+                extra = {
+                    'begin_event': event_chunk[0].timestamp,
+                    'end_event': event_chunk[len(event_chunk) - 1].timestamp
+                }
+                current_app.logger.warning(msg, extra=extra)
+                raise Exception()
```

Comment:
Since an exception is raised anyways, you can create a custom exception (like e.g. [MissingCommunityError](https://github.com/zenodo/zenodo/blob/9f0c975184f5f59e2aaf503134ceec8873b29a3e/zenodo/modules/deposit/errors.py#L64:L72)) and include the `msg` and `extra` dictionary